app/services/users/username_generator.rb

```diff
@@ -0,0 +1,46 @@
+module Users
+  class UsernameGenerator
+    def self.call(...)
+      new(...).call
+    end
+
+    def initialize(user)
+      @user = user
+    end
+
+    def call
+      username = from_auth_providers
+      return username if username.present? && !username_exists?(username)
+
+      3.times do
+        username = random_letters
+
+        break unless username_exists?(username)
+
+        username = nil
+      end
+
+      username
+    end
+
+    private
+
+    # @todo bit heavy check, maybe keep all usernames in redis?
+    def username_exists?(username)
+      User.exists?(username: username) ||
+        Organization.exists?(slug: username) ||
+        Page.exists?(slug: username) ||
+        Podcast.exists?(slug: username)
+    end
+
+    def from_auth_providers
+      provider = Authentication::Providers.username_fields.detect { |u| @user.public_send(u).present? }
```

Comment:
If there is no username, an exception will be raised, e.g. 
```
[4] pry(main)> Users::UsernameGenerator.call User.new
TypeError: nil is not a symbol nor a string
from /home/light/sites/dev.to/app/services/users/username_generator.rb:38:in `public_send'
```
We should handle a case when there is no corresponding "provider" username.